pkg/scheduler/algorithm/priorities/util/topologies.go

```diff
@@ -18,36 +18,91 @@ package util
 
 import (
 	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
-	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
 )
 
-// GetNamespacesFromPodAffinityTerm returns a set of names
+// GetNamespacesFromPodAffinityTerm returns a list of names
 // according to the namespaces indicated in podAffinityTerm.
 // If namespaces is empty it considers the given pod's namespace.
-func GetNamespacesFromPodAffinityTerm(pod *v1.Pod, podAffinityTerm *v1.PodAffinityTerm) sets.String {
-	names := sets.String{}
-	if len(podAffinityTerm.Namespaces) == 0 {
-		names.Insert(pod.Namespace)
-	} else {
-		names.Insert(podAffinityTerm.Namespaces...)
+func GetNamespacesFromPodAffinityTerm(pod *v1.Pod, podAffinityTerm *v1.PodAffinityTerm) []string {
+	if len(podAffinityTerm.Namespaces) > 0 {
+		return podAffinityTerm.Namespaces
 	}
-	return names
+	return []string{pod.Namespace}
 }
 
 // PodMatchesTermsNamespaceAndSelector returns true if the given <pod>
 // matches the namespace and selector defined by <affinityPod>`s <term>.
-func PodMatchesTermsNamespaceAndSelector(pod *v1.Pod, namespaces sets.String, selector labels.Selector) bool {
-	if !namespaces.Has(pod.Namespace) {
-		return false
+func PodMatchesTermsNamespaceAndSelector(pod *v1.Pod, namespaces []string, ps *metav1.LabelSelector) bool {
+	for _, ns := range namespaces {
+		if pod.Namespace == ns {
+			if LabelsMatchLabelSelector(labels.Set(pod.Labels), ps) {
+				return true
+			}
+		}
 	}
 
-	if !selector.Matches(labels.Set(pod.Labels)) {
+	return false
+}
+
+// LabelsMatchLabelSelector returns true if the given <labels> matches the labelSelector
+func LabelsMatchLabelSelector(ls labels.Labels, ps *metav1.LabelSelector) bool {
+	if ps == nil {
 		return false
 	}
+	for k, v := range ps.MatchLabels {
+		if !matchLabels(ls, k, v) {
+			return false
+		}
+	}
+	for _, expr := range ps.MatchExpressions {
+		if !matchExpression(ls, expr) {
+			return false
+		}
+	}
 	return true
 }
 
+func matchLabels(ls labels.Labels, key string, value string) bool {
+	if !ls.Has(key) {
+		return false
+	}
+	return value == ls.Get(key)
+}
+
+func expressionHasValue(expr metav1.LabelSelectorRequirement, value string) bool {
+	for i := range expr.Values {
```

Comment:
why not just do it like:
`for i,v := range expr.Values`

Reply:
done